DSharpPlus.Core/Rest/PollyRatelimitPolicy.cs

```diff
@@ -0,0 +1,152 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+using DSharpPlus.Caching.Abstractions;
+
+using Polly;
+
+namespace DSharpPlus.Core.Rest
+{
+    // main ratelimit handler
+    internal class PollyRatelimitPolicy : AsyncPolicy<HttpResponseMessage>
+    {
+        // keeps track of endpoint -> bucket hash mapping
+        private readonly ConcurrentDictionary<string, string> __endpoint_hash_mapping;
+
+        // stores one second, a very common thing, so we don't want to allocate it every time
+        private static readonly TimeSpan __one_second = TimeSpan.FromSeconds(1);
+
+        private const string GlobalBucketName = "DSharpPlus:Ratelimiting:global";
+
+        public PollyRatelimitPolicy()
+            => __endpoint_hash_mapping = new();
+
+
+        protected override async Task<HttpResponseMessage> ImplementationAsync
+        (
+            Func<Context, CancellationToken, Task<HttpResponseMessage>> action,
+            Context context,
+            CancellationToken cancellationToken,
+            bool continueOnCapturedContext = true
+        )
+        {
+            // ascertain a valid endpoint
+            if (!context.TryGetValue("endpoint", out object endpointObject) || endpointObject is not string endpoint)
+            {
+                throw new InvalidOperationException("No valid endpoint provided.");
+            }
+
+            //ascertain a valid cache.
+            if (!context.TryGetValue("cache", out object cacheObject) || cacheObject is not ICacheService cache)
```

Comment:
Consider taking this in from DI; users can also/should also be able to make un-limited requests.

Reply:
users should be able to make unlimited request, maybe use another context variable for that? also, this is indirectly taken in from DI through the rest client, which sets the cache.